jar-infer/jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/DefinitelyDerefedParams.java

```diff
@@ -98,15 +104,60 @@ private static void LOG(boolean cond, String tag, String msg) {
   Set<Integer> analyze() {
     // Get ExceptionPrunedCFG
     LOG(DEBUG, "DEBUG", "@ " + method.getSignature());
-    Set<Integer> derefedParamList = new HashSet<Integer>();
     prunedCFG = ExceptionPrunedCFG.make(cfg);
     // In case the only control flows are exceptional, simply return.
     if (prunedCFG.getNumberOfNodes() == 2
         && prunedCFG.containsNode(cfg.entry())
         && prunedCFG.containsNode(cfg.exit())
         && GraphUtil.countEdges(prunedCFG) == 0) {
-      return derefedParamList;
+      return new HashSet<>();
     }
+
+    // Get number of params and value number of first param
+    int numParam = ir.getSymbolTable().getNumberOfParameters();
+    int firstParamIndex =
+        method.isStatic() ? 1 : 2; // 1-indexed; v1 is 'this' for non-static methods
+
+    Set<Integer> derefedParamList;
+    if (USE_EXTENDED_APPROACH) {
+      derefedParamList = computeDerefParamList(numParam, firstParamIndex);
+    } else {
+      derefedParamList = computeDerefParamListUsingPDom(numParam, firstParamIndex);
+    }
+    return derefedParamList;
+  }
+
+  @NotNull
+  private Set<Integer> computeDerefParamList(int numParam, int firstParamIndex) {
+    Set<Integer> derefedParamList = new HashSet<>();
+    Map<ISSABasicBlock, Set<Integer>> blockToDerefSetMap = new HashMap<>();
+
+    // find which params are definitely non-null in each block
```

Comment:
nit, slightly clearer: "find which params are dereferenced (and thus non-null) in each block"

Reply:
actually, we test more things than just dereferencing, like precondition and assertions, hence I phrased it like this. And maybe in future we might add more checks which are not just deref too.

Reply:
Maybe "find which params are treated as being non-null inside each basic block", then? :)

And for the comment on the filter:

"For each param `p`, if no path exists from the entry block to the exit block which traverses only basic blocks which do not require  `p` being non-null (e.g. by dereferencing it), then mark `p` as `@NonNull`"

The main problem I have with "params are definitely non-null in each block" is that it's not clear that they are expected to be non-null on entry vs guaranteed to be non-null by the block.

Reply:
Yeah..makes sense